File: codes/vb/factorial.vb
Imports System.Collections.Generic ' Dictionary, KeyValuePair

Module Program
    Function FactorialV1(ByVal Number As Integer) As Integer
        If (Number < 0) Then Throw New Exception("Argument should be >= 0")
        If (Number = 0) Then Return 1
        Dim Result As Integer = 1
        Dim I = Number
        While (true)
            Result *= I
            I -= 1
            If (I <= 0) Then Exit While
        End While
        Return Result
    End Function

    Function FactorialV2(ByVal Number As Integer) As Integer
        If (Number < 0) Then Throw New Exception("Argument should be >= 0")
        If (Number = 0) Then Return 1
        Dim Result As Integer = 1
        Dim I = Number
        While (I >= 1)
            Result *= I
            I -= 1
        End While
        Return Result
    End Function

    Function FactorialV3(ByVal Number As Integer) As Integer
        If (Number < 0) Then Throw New Exception("Argument should be >= 0")
        If (Number = 0) Then Return 1
        Dim Result As Integer = 1
        For I As Integer = 5 To 1 Step -1
            Result *= I
        Next
        Return Result
    End Function

    Function FactorialV4(ByVal Number As Integer) As Integer
        If (Number < 0) Then Throw New Exception("Argument should be >= 0")
        If (Number = 0) Then Return 1
        Return (Number * FactorialV4(Number - 1))
    End Function

    Sub Main(Args As String())
        Console.WriteLine($"{Environment.NewLine}' Factorial(n) in Visual Basic (.NET)")

        Console.WriteLine("' using factorial function ""FactorialV1""")
        Console.WriteLine($"Factorial(5): {FactorialV1(5)}")
        ' Factorial(5): 120

        Console.WriteLine("' using factorial function ""FactorialV2""")
        Console.WriteLine($"Factorial(5): {FactorialV2(5)}")
        ' Factorial(5): 120

        Console.WriteLine("' using factorial function ""FactorialV3""")
        Console.WriteLine($"Factorial(5): {FactorialV3(5)}")
        ' Factorial(5): 120

        Console.WriteLine("' using factorial function ""FactorialV4""")
        Console.WriteLine($"Factorial(5): {FactorialV4(5)}")
        ' Factorial(5): 120
    End Sub
End Module

File: codes/scala/factorial.scala
println("\n// Factorial(n) in Scala")

def factorialV1(number: Int): Int = {
    if (number < 0) throw new Exception("Argument should be >= 0")
    if (number == 0) return 1
    var result = 1
    var i = number
    while (true) {
        result *= i
        if (i <= 1) return result
        i -= 1
    }
    return result
}

def factorialV2(number: Int): Int = {
    if (number < 0) throw new Exception("Argument should be >= 0")
    if (number == 0) return 1
    var result = 1
    var i = number
    while (i >= 1) {
        result *= i
        i -= 1
    }
    return result
}

def factorialV3(number: Int): Int = {
    if (number < 0) throw new Exception("Argument should be >= 0")
    if (number == 0) return 1
    var result = 1
    for (i <- (number to 1 by -1)) {
        result *= i
    }
    return result
}

def factorialV4(number: Int): Int = {
    if (number < 0) throw new Exception("Argument should be >= 0")
    if (number == 0) return 1
    return (number * factorialV4(number - 1))
}

println("// using factorial function \"factorialV1\"")
println(s"Factorial(5): ${factorialV1(5)}")
// Factorial(5): 120

println("// using factorial function \"factorialV2\"")
println(s"Factorial(5): ${factorialV2(5)}")
// Factorial(5): 120

println("// using factorial function \"factorialV3\"")
println(s"Factorial(5): ${factorialV3(5)}")
// Factorial(5): 120

println("// using factorial function \"factorialV4\"")
println(s"Factorial(5): ${factorialV4(5)}")
// Factorial(5): 120
